### ISSUE TITLE
bug - call system.platform() fails

### ISSUE THREAD

**Issue by JarbasAI** [MycroftAI/mycroft-core#715]
if i do 

```
import platform
print platform.system()
```

 works, but not if inside a skill, i get:

> Traceback (most recent call last):
>   File "/home/user/mycroft-core-base/mycroft/skills/core.py", line 106, in load_skill
>     skill = skill_module.create_skill()
>   File "/home/user/mycroft-skills/mycroft-facebook-skill/__init__.py", line 977, in create_skill
>     return FacebookSkill()
>   File "/home/user/mycroft-skills/mycroft-facebook-skill/__init__.py", line 760, in __init__
>     print platform.system()
>   File "/usr/lib/python2.7/platform.py", line 1333, in system
>      return uname()[0]
>    File "/usr/lib/python2.7/platform.py", line 1300, in uname
>      processor = _syscmd_uname('-p','')
>   File "/usr/lib/python2.7/platform.py", line 1035, in _syscmd_uname
>     rc = f.close()
> IOError: [Errno 10] No child processes

seems related to:
https://issues.apache.org/jira/browse/SPARK-1394?focusedCommentId=14046349&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-14046349

this breaks a few packages, so far got problems on selenium

**Comment by forslund** [comment]
If I recall correctly this is due to the handling of SIGCHLD, the current
handler interferes with the python internals.

I'll see if I can find a better way to clean up subprocesses.

Den 1 maj 2017 8:32 em skrev "JarbasAI" <notifications@github.com>:

**Comment by SoloVeniaASaludar** [comment]
Doing some tests about this issue...

1) it is true that signal(SIGCHLD,SIG_IGN) breaks platform.system() call. Platform still uses deprecated "os.popen".

2) After an statement as "subprocess.Popen(["uname","-p"])", a defunct process appears. It still on this state until garbage collection does a cleanup (for testing, it can be forced with "gc.collect()").

3) After a call as "x=subprocess.Popen(["uname","-p"])" a defunct process appears. It still on this state until x is free and garbage collection does a cleanup. A way to force it is:

```
x=subprocess.Popen(["uname","-p"]
del(x)

```
this one seems the better way to force the dead of the zombie child.


In conclusion: call to "signal" can be suppressed. If the skill needs immediate finish of dead child, use the code said in point 3.

My apologizes for suggesting a bad change.

**Comment by forslund** [comment]
Is there a risk that the `del(x)` will abort a program that's taking some time to execute?

Reading the web it seems to me that the best method is to use `x.wait()` to wait for termination when possible.

I wonder if there is a way to automatically clean up pending subprocess when restarting a skill or detach processes completely. Maybe this is what the `del(x)` will do?

**Comment by forslund** [comment]
http://stackoverflow.com/questions/16807603/python-non-blocking-non-defunct-process has an example for a PID recycling mechanism that cleans up defunct processes

**Comment by SoloVeniaASaludar** [comment]
My test results in:

a) if "del" is called before child end, it has no effect. Child will go to "defunct".

b) if "del" is called after child end, defunct child is finished. But, as you say, there are other ways to achieve this result: x.wait(), ... .

I do not see any way to terminate defunct child immediately or, at least, only a few seconds after its end, except if parent enters in a loop of check child termination.

**Comment by forslund** [comment]
Resolved via #754
